examples/small/lazyeval.makam

```diff
@@ -0,0 +1,196 @@
+
+eval (Heap_Body) Result
+    when heap.open Heap_Body (pfun XS Defs (lam X_E) => success) :-
+  eq Heap_Body Result.
+
+(*
+
+   Γ : e ⇓ Δ : λy.e'   Δ : e'[x/y] ⇓ Θ : z
+   ————————————————————————————————————————
+              Γ : e x ⇓ Θ : z
+
+*)
+
+eval (Heap0_Body0) (Heap2_Body2)
+  when heap.open Heap0_Body0 (pfun XS Defs (app E X) => success) :-
+
+  heap.open Heap0_Body0 (pfun XS Defs (app E X) =>
+    heap.apply Heap0_Fun XS Defs E),
+  eval Heap0_Fun Heap1_Body1,
+
+  heap.open Heap1_Body1 (pfun XS Defs (lam Y_E') => [E X]
+    heap.apply_needed Heap0_Body0 XS (app E X),
+    heap.apply Heap1_Body2 XS Defs (Y_E' X)),
+  eval Heap1_Body2 Heap2_Body2.
+
+
+(*
+
+  —————————————-
+  Γ : n ⇓ Γ : n
+
+*)
+
+eval (Heap_Body) Result
+    when heap.open Heap_Body (pfun XS Defs (intconst N) => success) :-
+  eq Result Heap_Body.
+
+(*
+
+   Γ : e1 ⇓ Δ : n1     Δ : e2 ⇓ Θ : n2    OP n1 n2 = n3
+   —————————————————————————————————————————————————————
+              Γ : binop OP e1 e2 ⇓ Θ : n3
+
+*)
+
+eval (Heap0_Body0) (Heap2_Result)
+  when heap.open Heap0_Body0 (pfun XS Defs (binop OP E1 E2) => success) :-
+
+  heap.open Heap0_Body0 (pfun XS Defs (binop OP E1 E2) =>
+    eq OP_ OP,
+    heap.apply Heap0_E1 XS Defs E1),
+  eval Heap0_E1 Heap1_V1,
+
+  heap.open Heap1_V1 (pfun XS Defs (intconst N1) => [OP E1 E2]
+    eq N1 N1_,
+    heap.apply_needed Heap0_Body0 XS (binop OP E1 E2),
+    heap.apply Heap1_E2 XS Defs E2),
+  eval Heap1_E2 Heap2_V2,
+
+  heap.open Heap2_V2 (pfun XS Defs (intconst N2) =>
+    eq N2 N2_),
+  
+  OP_ N1_ N2_ N3,
+
+  heap.open Heap2_V2 (pfun XS Defs V2 =>
+    heap.apply Heap2_Result XS Defs (intconst N3)).
+
+(*
+           Γ : e ⇓ Δ : z
+  —————————————-—————————————————————
+  (Γ, x |-> e) : x ⇓ (Δ, x |-> z) : z
+
+  here we differ in that we keep the binding for x alive in the premise
+
+*)
+
+eval (Heap0_Body0) (Heap2_Val)
+  when heap.open Heap0_Body0 (pfun XS Defs (var X) => success) :-
+
+  heap.open Heap0_Body0 (pfun XS Defs (var X) => [E]
+    list.index_of (pfun Y => eq Y X) XS N,
+    list.nth Defs N E,
+    heap.apply Heap0_E XS Defs E),
+
+  eval Heap0_E Heap1_Val,
+  heap.open Heap1_Val (pfun XS Defs Val => [Defs']
+    list.update_nth Defs N Val Defs',
+    heap.apply Heap2_Val XS Defs' Val).
+
+(*
+   Γ, x1 |-> e1 ... xn |-> en : e ⇓ Δ : z
+  —————————————-—————————————————————------
+  Γ : let x1 = e1 ... xn = en in e ⇓ Δ : z
+
+*)
+
+eval (Heap0_Body0) (Heap2_V)
+  when heap.open Heap0_Body0 (pfun XS Defs (let YS_Body) => success) :-
+
+  heap.open Heap0_Body0 (pfun XS XSDefs (let YS_E) =>
+    bindmany.open YS_E (pfun YS (YSDefs, E) => [XSYS XSYSDefs]
+      append XS YS XSYS,
+      append XSDefs YSDefs XSYSDefs,
+      heap.apply Heap1_E XSYS XSYSDefs E)),
+
+  eval Heap1_E Heap2_V.
+
+(* and now for some tests *)
+
+maintests : testsuite. %testsuite maintests.
+
+>> eval (heap (body ([], binop plus (intconst 10) (intconst 5)))) X ?
+>> Yes:
+>> X := heap (body (tuple [  ] (intconst 15))).
+
```

Comment:
I would add the following test, the other ones don't actually need `apply_needed` and work just fine with `apply`.

```suggestion

>> eval (heap (body ([], binop plus (let (bind "u" (fun u => body([intconst 6], var u)) )) (intconst 14)  ))) X ?
>> Yes:
>> X := heap (bind "u" (fun u => body (tuple [ (intconst 6) ] (intconst 20)))).
```

Reply:
Thanks!